### PR TITLE
Add a "Process" operation for messaging

### LINKED ISSUE #657: Add semantic conventions for "Process" spans for messaging scenarios
Currently, the messaging workgroup works on an https://github.com/open-telemetry/oteps/pull/220 which covers span structures in messaging scenarios.

The group decided to focus on a consistent set of conventions that can be applied across _all_ messaging scenarios, which resulted into a proposal for conventions for "Publish", "Create", "Deliver", "Receive", and "Settle" operations, as those share common characteristics across all messaging scenarios. The same can't be said of "Process" operations, which can vary considerable depending on the individual use-case.

However, as interest was expressed from many sides to also achieve some consistency for the instrumentation of "Process" operations, it is necessary to provide conventions for "Process" operations as an addition to what's already in https://github.com/open-telemetry/oteps/pull/220.

This work doesn't block merging https://github.com/open-telemetry/oteps/pull/220.

### PR DESCRIPTION
**Pull request by pyohannes**
Fixes #657

## Changes

Interest was expressed from many sides to also achieve some consistency for the instrumentation of "Process" operations. For this, it is necessary to provide conventions for "Process" operations.

During the discussion of this PR, it was decided to _merge "Deliver" and "Process" operations_. See [this thread](https://github.com/open-telemetry/semantic-conventions/pull/700#discussion_r1481887367) for the reasoning behind this decision.

This PR adds semantic conventions for "Process" spans and metrics. This will not be part of the first iteration of stable messaging semantic conventions, but should serve for uncovering any related gaps we have in parts of conventions that are to be stabilized.

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] Change log entry added, according to the guidelines in [When to add a changelog entry](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#when-to-add-a-changelog-entry).
  * If your PR does not need a change log, start the PR title with `[chore]`
* [ ] ~~[schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.~~

### CODE REVIEW

**Comment by pyohannes** [comment]
@open-telemetry/semconv-messaging-approvers

File: docs/messaging/messaging-spans.md
Comment (lmolkova) [inline docs/messaging/messaging-spans.md:276]: It's hard to distinguish deliver and process:

- For example, Azure messaging client libraries have processing client (e.g. [EventHubProcessingClient](https://learn.microsoft.com/java/api/com.azure.messaging.eventhubs.eventprocessorclient?view=azure-java-stable) with API like `processEvents`  callback. 
- Azure also supports HTTP-push like [EventGrid](https://learn.microsoft.com/en-us/azure/event-grid/push-delivery-overview) along with the extension supporting Azure Functions triggers. 
- Such processor clients are designed is that if processing fails, messages should be redelivered. I.e. settlement is expected to happen within processor callback (but it's effectively up to the user in common case).

Similarly, [AWS SNS](https://docs.aws.amazon.com/sns/latest/dg/sns-message-delivery-retries.html#creating-delivery-policy) and [GCP PubSub](https://cloud.google.com/pubsub/docs/push) supports HTTP-push with redeliveries if request fails.

The typical [RabbitMQ](https://www.rabbitmq.com/confirms.html#acknowledgement-modes) use-case is also to settle message within consume/deliver callback.

So, deliver operation (at least for the common case)  is expected to also settle a message within it's scope and messaging brokers are designed to retry delivery if delivery fails or is negatively acked.

Also, deliver operations are those which can be automatically instrumented (by SDKs, FaaS extensions, etc).

Given all of this, I find it hard to draw a line between processing and delivery.

So, I think we should either:
- merge them together and maybe allow using different terminology
- clearly explain the difference  (e.g. processing is an application-level span which traces individual messages within a batch. We don't expect it to be created automatically, but we recommend (lowercase) everyone creating it to use the following semantics ...)

File: docs/messaging/messaging-spans.md
Comment (lmolkova) [inline docs/messaging/messaging-spans.md:210]: should we also specify a kind (Server or Internal)?

File: docs/messaging/messaging-spans.md
Comment (joaopgrassi) [reply docs/messaging/messaging-spans.md:276]: I think the case where we would have a clear distinction is the pull-based scenario, like you mentioned:

> or a span that traces processing when pull-based receive is used - in this case it's semantics SHOULD match deliver operation. Sometimes it can be created automatically (e.g. by integration libraries such as [spring-kafka](https://www.baeldung.com/spring-kafka#2-consuming-messages))

So I think then it makes sense to have the process semantics, but make it clear that case.

Comment (pyohannes) [reply docs/messaging/messaging-spans.md:276]: Discussed in the workgroup meeting: we will change this PR to merge "process" and "deliver" spans and metrics.

Comment (pyohannes) [reply docs/messaging/messaging-spans.md]: This PR now unifies "deliver" and "process" operations (and metrics).

@open-telemetry/semconv-messaging-approvers, please have a look.

File: docs/messaging/messaging-spans.md
Comment (pyohannes) [reply docs/messaging/messaging-spans.md]: Added. After the operation is unified with "deliver", it needs to be `CONSUMER` for push-based scenarios.

**Review by lmolkova — APPROVED** [review]
LGTM!

File: docs/messaging/messaging-spans.md
Comment (hongalex) [inline docs/messaging/messaging-spans.md:255]: So I agree with this approach, that the "process" span should only encapsulate the actual processing of the message by the callback in push based scenarios. As mentioned in the meeting, Google Pub/Sub client libraries do additional processing / transformations on the message prior to calling the user callback that we're interested in tracing as internal spans. In this case, the pre-processing spans would link to the "create" span, rather than the "process" span. Our process span would actually be a child of the pre-process span. I'm wondering if that conflicts with the diagram below (where the process span is linked to the create span).

Here's an example of what we have so far. The screenshot shows the trace of a single message produced and consumed, by a topic named `otel-topic` and consumer named `otel-sub`.
image

File: docs/messaging/messaging-spans.md
Comment (hongalex) [inline docs/messaging/messaging-spans.md:487]: I think this might have been discussed before, but what is the reasoning that these are links instead of parent/child relationships? My understanding was that links are needed for scenarios where a span might have multiple parents but this doesn't seem to be the case. Should we offer that either (links or parent/child) can be used?

File: docs/messaging/messaging-spans.md
Comment (pyohannes) [reply docs/messaging/messaging-spans.md:255]: That looks good, you have the required "Create" and "Process" spans present.

The "Process" span should definitely link to the "Create" span. It can still be child of the "subscribe" span.

For naming the "Process" span, we recommend `<destination name> <operation name>`, which in your case should be `otel-topic process` I think?

File: docs/messaging/messaging-spans.md
Comment (lmolkova) [reply docs/messaging/messaging-spans.md]: it's no longer relevant after process and deliver were merged

File: docs/messaging/messaging-spans.md
Comment (pyohannes) [reply docs/messaging/messaging-spans.md:487]: We require links to be present, as that's the only reliable correlation mechanism that can work consistently across all messaging scenarios.

Parent/child relationships can be added in addition to links where feasible, but aren't required by the conventions. For the sake of keeping examples minimal, I left it out there.

**Review by hongalex — APPROVED** [review]

File: docs/messaging/messaging-spans.md
Comment (hongalex) [reply docs/messaging/messaging-spans.md:487]: Yeah that sounds good. Thanks for the clarification.

File: docs/messaging/messaging-spans.md
Comment (hongalex) [reply docs/messaging/messaging-spans.md:255]: Yeah adding additional links sounds fine.

**Review by carlosalberto — APPROVED** [review]

**Comment by carlosalberto** [comment]
LGTM. As mentioned before, an example of Receive +  Process would be useful, even if more of a specific case - but definitely not blocking.

**Review by joaopgrassi — APPROVED** [review]
Looks good to me. 

Just a nit: We are not just adding `Process` but we are in fact, replacing `Deliver` with Process`. Maybe some context on why this change can help reviewers.

**Comment by pyohannes** [comment]
> Just a nit: We are not just adding Process but we are in fact, replacing Deliver with Process`. Maybe some context on why this change can help reviewers.

@joaopgrassi I added a note about this to the PR description, referencing the relevant thread.

**Pull request merged**
